src/python/bow/bow.py

```diff
@@ -0,0 +1,224 @@
+        newHeight = (height * kDefaultWidth) / width
+        image = cv2.resize(image, (kDefaultWidth, int(newHeight)))
+        print("Resized image from", height, width, "to", image.shape)
+    return image
+
+
+def extractSiftsFromImage(imageFile):
+    """Extracts SIFT features from an image
+
+    Args:
+        imageFile (Path): path to the image file
+
+    Returns:
+        list(list(int)): array of descriptors NxD
+    """
+    # Extracts features from an image
+    image = cv2.imread(imageFile.as_posix(), cv2.IMREAD_GRAYSCALE)
+    image = rescaleImageIfNeeded(image)
+    sift = cv2.SIFT_create()
+    keypoints, descriptors = sift.detectAndCompute(image, None)
+    return descriptors
+
+
+def computeIDF(descriptorsPerImage, clusters):
+    """Compute inverse document frequence (IDF). Here means in how many images does the word occur.
+
+    Args:
+        descriptorsByImages (list(list(1xD)): List of descriptors per image
+        clusters (np.array): CxD array of clusters (words)
+    Returns:
+        np.array: Cx1 occurence of clusters/words in images
+    """
+    clusterOccurenceInImages = [set() for index in range(clusters.shape[0])]
+    N = len(descriptorsPerImage)
+    clustersTree = KDTree(clusters)
+    for imageId in range(len(descriptorsPerImage)):
+        dist, nearestClusters = clustersTree.query(descriptorsPerImage[imageId], k=1)
+        for clusterId in nearestClusters.squeeze():
+            if clusterId < 0 or clusterId >= clusters.shape[0]:
+                print("Error: cluster ids outside bounds")
+                continue
+            clusterOccurenceInImages[clusterId].add(imageId)
+
+    # reweight by number of images
+    clusterOccurence = [0] * clusters.shape[0]
+    for clusterId in range(len(clusterOccurenceInImages)):
+        if len(clusterOccurenceInImages[clusterId]) <= 0:
+            print("WARNING: word", clusterId, "is not represented in any image")
+            continue
+        clusterOccurence[clusterId] = N / len(clusterOccurenceInImages[clusterId])
+    return np.array(clusterOccurence)
+
+
+def trainVocabulary(imageFiles, outputFile=""):
+    """train vocabulary from given image paths
+
+    Args:
+        imageFiles (list(Path)): paths to images
+    """
+    descriptorsPerImage = []
+    for imageFile in imageFiles:
+        sifts = extractSiftsFromImage(imageFile)
+        descriptorsPerImage.append(sifts)
+
+    # flatten the descriptors list
+    descriptors = [
+        descriptor for descriptors in descriptorsPerImage for descriptor in descriptors
+    ]
+    descriptors = np.array(descriptors)
+
+    descriptorsNormalized = preprocessing.normalize(descriptors)
+    kmeans = KMeans(n_clusters=kDefaultClusterSize, random_state=0, n_init="auto")
+    kmeans.fit(descriptorsNormalized)
+    words = kmeans.cluster_centers_
+
+    idfs = computeIDF(descriptorsPerImage, words)
+
+    plt.bar(range(0, len(idfs)), idfs)
+    plt.savefig("idf_" + str(kDefaultClusterSize) + ".png")
```

Comment:
nit: Do you want to always save these images? Or should it be hidden behind some flag?

Reply:
Neh, not really important. Will just remove it and reduce the dependencies list.